Core/src/AmbiguityResolution/AthenaAmbiguityResolution.cpp

```diff
@@ -0,0 +1,216 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2023 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "Acts/AmbiguityResolution/AthenaAmbiguityResolution.hpp"
+
+#include "Acts/EventData/Measurement.hpp"
+#include "Acts/EventData/SourceLink.hpp"
+
+#include <stdexcept>
+
+std::vector<std::size_t> Acts::AthenaAmbiguityResolution::getCleanedOutTracks(
+    std::vector<double> trackScore,
+    std::vector<std::map<std::size_t, Counter>>& counterMaps,
+    std::vector<std::vector<std::tuple<std::size_t, std::size_t, bool>>>
+        measurementsPerTrack) const {
+  std::vector<std::size_t> cleanTracks;
+
+  ACTS_INFO("Cleaning tracks");
+
+  if (trackScore.size() != measurementsPerTrack.size()) {
+    throw std::invalid_argument(
+        "Track score and measurementsPerTrack size mismatch");
+  }
+
+  std::size_t numberOfTracks = measurementsPerTrack.size();
+  ACTS_DEBUG("Number of tracks: " << numberOfTracks);
+
+  boost::container::flat_map<std::size_t,
+                             boost::container::flat_set<std::size_t>>
+      tracksPerMeasurement;
+
+  for (std::size_t iTrack = 0; iTrack < numberOfTracks; ++iTrack) {
+    if (trackScore[iTrack] <= 0) {
+      measurementsPerTrack[iTrack].clear();
+      continue;
+    }
+    for (auto measurements_tuples : measurementsPerTrack[iTrack]) {
+      auto iMeasurement = std::get<0>(measurements_tuples);
+      tracksPerMeasurement[iMeasurement].insert(iTrack);
+    }
+  }
+
+  enum TsosTypes {
+    // A measurement not yet used in any other track
+    UnusedHit = 1,
+    // A measurement shared with another track
+    SharedHit = 2,
+    // A hit that needs to be removed from the track
+    RejectedHit = 3,
+    // an outlier, to be copied in case
+    Outlier = 4,
+    // other TSOS types to be copied in case
+    OtherTsos = 5
+  };
```

Comment:
What is a Tsos ? either remove the abbreviation or explain

Reply:
I will try to make a better sounding name.

Reply:
It is not so much about it sounding better it is about clarity